### PR TITLE
Allow for instant widget content updates

### PR DESCRIPTION
**Pull request by tMaxx**

### CODE REVIEW

File: cogboard-webapp/src/helpers/index.js [hunk 1]
```diff
@@ -41,3 +43,17 @@ export const parseYupErrors = errors => {
...
+    body: JSON.stringify(data),
+    headers: {
+      'Content-Type': 'application/json'
+    }
+  };
+
+  return fetch('/api/widget/contentUpdate', postConfig)
+    .then(checkResponseStatus)
+    .then(response => response.json());
+};
```
Comment (tMaxx) [inline cogboard-webapp/src/helpers/index.js]: Should probably move to `utils`. Which brings me to the question - why do we have `utils` and `helpers`? As far as I can tell, there is little to no difference, so both directories could potentially get merged.

File: cogboard-app/src/main/kotlin/com/cognifide/cogboard/CogboardConstants.kt [hunk 2]
```diff
@@ -47,6 +48,7 @@ class CogboardConstants {
 
         const val EVENT_SEND_MESSAGE_TO_WEBSOCKET = "cogboard.websocket.message"
         const val EVENT_SAVE_BOARDS_CONFIG = "cogboard.config.boards.save"
+        const val EVENT_CONTENT_UPDATE_WIDGET_CONFIG = "cogboard.config.widget.contentUpdate"
```
Comment (szymon-owczarzak) [inline cogboard-app/src/main/kotlin/com/cognifide/cogboard/CogboardConstants.kt]: i think better name would be `EVENT_UPDATE_WIDGET_CONTENT `

File: cogboard-app/src/main/kotlin/com/cognifide/cogboard/widget/BaseWidget.kt [hunk 3]
```diff
@@ -63,11 +65,11 @@ abstract class BaseWidget(
      */
     fun sendConfigurationError(cause: String = "") {
         send(JsonObject()
-                .put(CC.PROP_STATUS, Widget.Status.ERROR_CONFIGURATION)
                 .put(CC.PROP_CONTENT,
                         JsonObject()
                                 .put(CC.PROP_ERROR_MESSAGE, "Configuration Error")
                                 .put(CC.PROP_ERROR_CAUSE, cause)
+                                .put(CC.PROP_WIDGET_STATUS, Widget.Status.ERROR_CONFIGURATION)
```
Comment (szymon-owczarzak) [inline cogboard-app/src/main/kotlin/com/cognifide/cogboard/widget/BaseWidget.kt]: Yes status as a part of content

File: cogboard-app/src/main/kotlin/com/cognifide/cogboard/CogboardConstants.kt (hunk 2 above)
Comment (tMaxx) [reply cogboard-app/src/main/kotlin/com/cognifide/cogboard/CogboardConstants.kt]: Agreed, fixed.

File: cogboard-app/src/main/kotlin/com/cognifide/cogboard/widget/type/TextWidget.kt [hunk 4]
```diff
@@ -6,6 +6,13 @@ import io.vertx.core.json.JsonObject
 
 class TextWidget(vertx: Vertx, config: JsonObject) : BaseWidget(vertx, config) {
 
+    init {
+        createDynamicChangeSubscriber()
+            .handler {
+                send(it.body())
+            }
+    }
+
```
[the remark this answers is not in the document]
Comment (tMaxx) [reply cogboard-app/src/main/kotlin/com/cognifide/cogboard/widget/type/TextWidget.kt]: Mystery solved, to be removed.

File: cogboard-app/src/main/kotlin/com/cognifide/cogboard/widget/type/CheckboxWidget.kt [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+            .handler { send(it.body()) }
+    }
+
+    override fun updateState() {
+        updateStateByCopingPropsToContent(PROPS)
+    }
+
+    companion object {
+        val PROPS = setOf("widgetStatus")
+    }
```
[the remark this answers is not in the document]
Comment (tMaxx) [reply cogboard-app/src/main/kotlin/com/cognifide/cogboard/widget/type/CheckboxWidget.kt]: Not needed, removed.

File: cogboard-webapp/src/helpers/index.js (hunk 1 above)
Comment (tMaxx) [reply cogboard-webapp/src/helpers/index.js]: Waiting for Marek's input on this one

File: cogboard-webapp/src/components/Widget/index.js [hunk 6]
```diff
@@ -109,7 +110,7 @@ const Widget = ({ id, index }) => {
   return (
     <>
       <StyledCard
-        status={status}
```
Comment (szymon-owczarzak) [inline cogboard-webapp/src/components/Widget/index.js]: have you removed this constant ? `status`

Comment (tMaxx) [reply cogboard-webapp/src/components/Widget/index.js]: Good catch, fix will be in the next commit.

File: cogboard-webapp/src/helpers/index.js (hunk 1 above)
Comment (tMaxx) [reply cogboard-webapp/src/helpers/index.js]: Agreed on and modified per mutual observations.

**Pull request merged**